CookieboxCalibration: Add recipe test, allow deconvolution-calibration embedding, simplify normalization

daniloefl opened the pull request:
As promised, a test has been added to test changes in the CookieboxCalibration code, using an existing commissioning run (and therefore this requires GPFS, so it runs only if GPFS is available. That should be triggered in https://git.xfel.eu/dataAnalysis/extra-tests

In addition, previous code required some manual alignment if one wanted to use the `TOFAnalogResponse` and the `CookieboxCalibration`. Now, if one provides a dictionary of `TOFAnalogResponse` objects in the `setup` of `CookieboxCalibration`, this alignment is not needed, as the calibration fit is done on the deconvolved data directly (although this is slower -- still, it is the correct thing to do).

The normalization code has also been simplified to contain only the `normalization` factor (the other was not needed).

Finally, since the `_tof_response` variable may be `None`, depending or not on whether this is desired, the base class used for persistence has been adapted to cope with it.

philsmt commented on tests/mockdata/xgm.py:
```diff
@@ -56,3 +56,40 @@ class XGMReduced(XGMD):
         ("location", None, "XGM.3356.SQS")
     ]
+
+class XGMWithData(DeviceBase):
```
I didn't compare the exact keys, but would it make sense to inherit one of the above to save some definitions, and just overwrite `write_instrument()`?

daniloefl replied:
Indeed, except for the shape of the instrument array, which I tried to keep as 1D to avoid extra complications, but I can just override that in the daughter class.

philsmt commented on tests/test_recipes_cookiebox.py:
```diff
@@ -2,13 +2,277 @@
 import numpy as np
 import pandas as pd
 import xarray as xr
+import os
```
Are you using this import?

philsmt reviewed (commented):
That's roughly what I had in mind when going for entirely mockdata. I'm not all to familiar with the internal details, is this testing the code sufficiently in your opinion?

One might be able to save some space through integrating this run with others, but it doesn't seem terribly at the moment with a single real channel and one board.

daniloefl commented:
I have difficulties thinking of validation methods for the code that do more than this. I am not sure if it tests the code sufficiently, but the only thing I can compare with an `assert` and compare with the input are the calibration constants.

I tried to improve a bit the coverage, if that is what you had in mind:
- I noticed that the call to `calc_mean` is not shown in the coverage, but it is executed. The issue might be that I call it in parallel using a `ProcessPoolExecutor`. I added a test (and some code) to allow disabling that execution path, just so I can test it.
- I also added tests checking that the diagnostics plots work (which does not mean they plot good things, but if the `assert` passed, anything that does not crash is assumed to be intended, as we cannot easily validate plots).
- I also added tests for the deconvolution code (which in this commit has been integrated with the calibration to avoid misalignments -- although that is slower). This included only the analog deconvolution, and not the counting-mode based one.
- As I am skeptical that there is a use for the counting-mode instrumental response-based deconvolution, I have removed it, instead of writing a test for it. I tried using it in the past, but I honestly see no point to it, since the difference due to analog mode are too large.
- I removed a legacy function that the code coverage found as not being tested: it was not tested, because it is never used anywhere ...

Other things to test include some execution paths that are only triggered in case some arguments are given.

daniloefl commented:
To @JamesWrigley question: I added the `dump_etof_calibration_example.py` file, which reads the output of the calibration in an existing run and dumps some code to produce it. I was afraid of uploading the data itself in the repository, which is why I transformed it in code.

Nevertheless, the H5 file that I used to produce the constants is 2.4 MB. I could just read from there if it is ok to upload the H5 file itself.

daniloefl commented:
To avoid publishing the H5 file here, I added the code to read the run and produce the example constants from scratch (ie: from the proposal and run -- and other runs could be used instead to obtain the same effect). No H5 file is needed, therefore.

daniloefl commented on tests/test_recipes_cookiebox.py:
```diff
@@ -2,13 +2,407 @@
+                    stop_roi=320,
+                    parallel=False,
+    )
+    cal.setup(run=mock_sqs_etof_calibration_run, energy_axis=energy_axis, tof_settings=tof_channel,
```
Added an assert.

daniloefl commented on tests/test_recipes_cookiebox.py:
```diff
@@ -2,13 +2,407 @@
+    spectrum = cal_read.calibrate(data, subtract_offset=True)
+
+    # now call simpler function to do both at once
+    spectrum = cal_read.apply(mock_sqs_etof_calibration_run.select_trains(np.s_[10:20]))
```
Not those, but some others, yes.

daniloefl commented on tests/utils/dump_etof_calibration_example.py:
```diff
@@ -0,0 +1,120 @@
+import numpy as np
```
Agreed. I moved it.

daniloefl commented:
> Thanks, thanks for giving it a try entirely based on mockdata.
> 
> There are a few nitpicks left about minimizing random files lying around, and whether you could assert more.
> 
> I was wondering how this interacts with the code in `offline_tests`?

It does not at all. It was added before the mock data tests, so now it does not seem like the offline tests are necessary, so I removed them.
